samtranslator/model/sam_resources.py

```diff
@@ -2300,9 +2301,43 @@ def _construct_ddb_datasources(
                 cfn_datasource.Description = cast(PassThrough, ddb_datasource.Description)
 
             cfn_datasource.DynamoDBConfig = self._parse_ddb_config(ddb_datasource)
-            datasources.append(cfn_datasource)
 
-        return datasources
+            if ddb_datasource.ServiceRoleArn:
+                cfn_datasource.ServiceRoleArn = cast(PassThrough, ddb_datasource.ServiceRoleArn)
+            else:
+                # If the user doesn't have their own role, then we will create for them if TableArn and Permissions is defined.
```

Comment:
This is a good candidate for its own separate function that early-returns, avoiding the deeply nested (hence typically hard-to-read) blocks of code.